lib/resources/oidc.js

```diff
@@ -0,0 +1,184 @@
+      res.clearCookie(CODE_VERIFIER_COOKIE, callbackCookieProps);
+      res.clearCookie(NEXT_COOKIE,          callbackCookieProps); // eslint-disable-line no-multi-spaces
+
+      const client = await getClient();
+
+      const params = client.callbackParams(req);
+      const tokenSet = await client.callback(getRedirectUri(), params, { code_verifier });
+
+      const { access_token } = tokenSet;
+
+      const userinfo = await client.userinfo(access_token);
+
+      const { email, email_verified } = userinfo;
+      if (!email) {
+        // eslint-disable-next-line quotes
+        container.Sentry.captureException(new Error(`Required claim not provided in UserInfo Response: 'email'`));
+        return errorToFrontend(req, res, 'email-claim-not-provided');
+      }
+      if (!email_verified) return errorToFrontend(req, res, 'email-not-verified'); // eslint-disable-line camelcase
+
+      const user = await getUserByEmail(container, email);
+      if (!user) return errorToFrontend(req, res, 'auth-ok-user-not-found');
+
+      await initSession(container, req, res, user);
+
+      // This redirect would be ideal, but breaks `SameSite: Secure` cookies.
+      // return res.redirect('/');
+      // Instead, we need to render a page and then "browse" from that page to the normal frontend:
+
+      const nextPath = safeNextPathFrom(next);
+
+      // id=cl only set for playwright. Why can't it locate this anchor in any other way?
+      return {
+        head: html`<meta http-equiv="refresh" content="0; url=${nextPath}">`,
+        body: html`
+          <h1>Authentication Successful</h1>
+          <div><a href="${nextPath}" id="cl">Continue to ODK Central</a></div>
+        `,
+      };
+    } catch (err) {
+      container.Sentry.captureException(err);
+      // hack to override the defaultErrorWriter TODO should be fixed elsewhere
+      err.isProblem = true;
+      throw err;
+    }
+  }));
+};
+
+function errorToFrontend(req, res, errorCode) {
+  const loginUrl = new URL('/#/login', envDomain);
+
+  loginUrl.searchParams.append('oidcError', errorCode);
+
+  const next = req.cookies[NEXT_COOKIE];
+  if (next && !Array.isArray(next)) loginUrl.searchParams.append('next', next);
+
+  // Append query string manually, because Central Frontend expects search/hash
+  // in the wrong order (Vue hash-based routing).
+  const redirectUrl = envDomain + loginUrl.pathname + loginUrl.hash + loginUrl.search;
+
+  redirect(307, redirectUrl);
+}
+
+async function getUserByEmail({ Users }, email) {
+  const userOption = await Users.getByEmail(email);
+  if (!userOption.isDefined()) return;
+
+  const user = userOption.get();
+
+  return user;
+}
+
+async function initSession(container, req, res, user) {
+  const applySession = await createUserSession(container, req.headers, user);
+  applySession(req, res);
+}
+
+// Logic adapted from `login.vue` in frontend.
+function safeNextPathFrom(next) {
```

Comment:
this can be moved to `lib/util/http` or `lib/util/util`

+ can you please add a unit test as well